Refine the warnings about incompatible linter options

MichaReiser opened the pull request:
## Summary

Avoid warning about incompatible rules except if their configuration directly conflicts with the formatter. This should reduce the noise and potentially the need for https://github.com/astral-sh/ruff/issues/8175 and https://github.com/astral-sh/ruff/issues/8185

I also extended the rule and option documentation to mention any potential formatter incompatibilities or whether they're redundant when using the formatter. 

* `LineTooLong`: This is a use case we explicitly want to support. Don't warn about it
* `TabIndentation`, `IndentWithSpaces`: Only warn if `indent-style="tab"`
* `IndentationWithInvalidMultiple`, `IndentationWithInvalidMultipleComment`: Only warn if `indent-width != 4`
* `OverIndented`: Don't warn, but mention that the rule is redundant
* `BadQuotesInlineString`: Warn if quote setting is different from `format.quote-style`
* `BadQuotesMultilineString`, `BadQuotesDocstring`: Warn if `quote != "double"`

## Test Plan

I added a new integration test for the default configuration with `ALL`. `ruff format` now only shows two incompatible rules, which feels more reasonable.

MichaReiser commented:
Current dependencies on/for this PR:
* main
  * **PR #8192** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8192" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #8196** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8196" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/8196?utm_source=stack-comment).

MichaReiser commented on crates/ruff_cli/src/commands/format.rs:
```diff
@@ -685,31 +685,79 @@ pub(super) fn warn_incompatible_formatter_settings(
-            Rule::IndentWithSpaces,
+        for rule in [
+            // The formatter might collapse implicit string concatenation on a single line.
             Rule::SingleLineImplicitStringConcatenation,
```
I'm undecided if we should warn about this rule. The formatter might collapse implicit string concatenated lines which then triggers this rule. However, it might be something users want to fix manually.

yakMM replied:
If the formatter collapses the strings, I would love to know about it to fix it afterward.

yakMM commented:
Looks great, most of my warnings are gone!

MichaReiser commented:
[yakMM](https://github.com/yakMM) [wrote](https://github.com/astral-sh/ruff/issues/8185#issuecomment-1778845285)

> Tested on my project, only have a warning about ISC001 and isort.split-on-trailing-comma left. I will reply to your comment about this one directly in the PR.

> Don't know if I should open a new issue, but I feel like the warning about isort.split-on-trailing-comma=false is underserved since I already have isort.force-single-line=true, so that isort.split-on-trailing-comma=false will never be applied. I don't know how complicated it is to keep making exceptions to the warning on the implementation side.

charliermarsh commented on crates/ruff_cli/src/commands/format.rs:
```diff
@@ -700,31 +700,79 @@ pub(super) fn warn_incompatible_formatter_settings(
-            Rule::LineTooLong,
-            Rule::TabIndentation,
-            Rule::IndentationWithInvalidMultiple,
-            Rule::IndentationWithInvalidMultipleComment,
```
I these think these are redundant when using tabs in the formatter.

charliermarsh commented on crates/ruff_cli/src/commands/format.rs:
````diff
@@ -700,53 +700,105 @@ pub(super) fn warn_incompatible_formatter_settings(
+            //     pass
+            // ```
             Rule::MissingTrailingComma,
-            Rule::ProhibitedTrailingComma,
````
This was removed because the formatter will never output code that triggers it? Is that right?

charliermarsh replied:
Yes, I believe this is the case.
